mesonbuild/interpreter.py

```diff
@@ -3616,8 +3662,32 @@ def dependency_fallback(self, name, display_name, kwargs):
             'required': required,
         }
         self.do_subproject(dirname, 'meson', sp_kwargs)
+        self.override_dependencies_from_wrap(dirname, kwargs)
         return self.get_subproject_dep(name, display_name, dirname, varname, kwargs)
 
+    def override_dependencies_from_wrap(self, dirname, kwargs):
+        # Make sure to override every dependencies this wrap provides in the
+        # case the subproject did not call meson.override_dependency().
+        # This ensures that if we fallbacked for one dependency, any other
+        # dependency provided by this subproject won't be picked from system.
+        # Also verify consistency in the case the subproject overridden the
+        # dependency with a different object than the variable from wrap file.
+        for_machine = self.machine_from_native_kwarg(kwargs)
+        provide = self.environment.wrap_resolver.get_provide(dirname)
+        for depname, varname in provide.items():
+            if not varname:
```

Comment:
shouldn't this be verified too? i.e. that an override was done? Might make things awkward if options are involved. But maybe it needs to be marked as "permanently not found"?

Reply:
Hmm, I'm wondering if this should be considered a bug in the subproject, it should override the dep with not-found in the case it's disabled by option, maybe? This feels like really corner case, for now I'll override with not-found and print a warning about it, what do you think?

Reply:
Yes, i think overriding with not-found might be ok here. I think doing not-found overrides in the subproject for each disabled option can get to be quite a bit of boilerplate. But if we do this automatically here, it will be hard to change that later.
Maybe from the UI perspective it would be nice to override with a poisoned dependency that prints "subproject X failed to provide promised dependency, maybe an option is missing" on first use.